File: frontend/src/services/profileService.ts
import api from './api'

export interface UserProfile {
  id: number
  name: string
  email: string
  role: string
}

export const profileService = {
  async getProfile(): Promise<UserProfile> {
    const { data } = await api.get<UserProfile>('/users/profile')
    return data
  },

  async updateProfile(name: string): Promise<UserProfile> {
    const { data } = await api.put<UserProfile>('/users/profile', { name })
    return data
  },

  async changePassword(currentPassword: string, newPassword: string): Promise<void> {
    await api.put('/users/profile/password', { currentPassword, newPassword })
  },
}

File: src/main/java/com/tickets/service/ProfileService.java
package com.tickets.service;

import com.tickets.dto.UserDTO;
import com.tickets.entity.User;
import com.tickets.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class ProfileService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public ProfileService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public UserDTO getProfile(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return new UserDTO(user.getId(), user.getName(), user.getEmail(), user.getRole());
    }

    public UserDTO updateProfile(String email, String newName) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        user.setName(newName);
        User saved = userRepository.save(user);

        return new UserDTO(saved.getId(), saved.getName(), saved.getEmail(), saved.getRole());
    }

    public void changePassword(String email, String currentPassword, String newPassword) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new RuntimeException("Current password is incorrect");
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
}
